### PR TITLE
Add AdjustedPartitionsRDD

### PR DESCRIPTION
**Pull request by tpoterba**
This permits shuffle-free load of files where
 multiallelics are split across partitions.

### CODE REVIEW

**Comment by tpoterba** [comment]
@danking addressed a few outstanding issues, added a new test for coercion of unsorted-within-partitions RDDs.

File: src/main/scala/org/broadinstitute/hail/sparkextras/AdjustedPartitionsRDD.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.spark.rdd.RDD
+import org.apache.spark.{Dependency, NarrowDependency, Partition, TaskContext}
+
+import scala.reflect.ClassTag
+
+case class AdjustedPartitionsRDDPartition[T](index: Int, parent: Partition, adj: Adjustment[T]) extends Partition
+
+case class Adjustment[T](originalIndex: Int,
+  drop: Option[Iterator[T] => Iterator[T]],
+  take: Seq[(Int, Iterator[T] => Iterator[T])])
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/AdjustedPartitionsRDD.scala]: Make this more general: just have a list of partitions to take from with filters.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 2]
```diff
@@ -87,31 +87,33 @@ object OrderedRDD {
 
     val sortedKeyInfo = keyInfo.sortBy(_.min)
     val partitionsSorted = sortedKeyInfo.zip(sortedKeyInfo.tail).forall { case (p, pnext) =>
-      val r = p.max < pnext.min
+      val r = p.max <= pnext.min
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: Should we relax this?  Once you have to stream through the previous partition, it doesn't really matter if you take a little or a lot.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 3]
```diff
@@ -122,9 +124,67 @@ object OrderedRDD {
...
+      val i = it.next()
+      val thisP = sortedKeyInfo(i)
+      val min = thisP.min
+      val max = thisP.max
+
+      indicesBuilder += i
+
+      def overlaps(info: PartitionKeyInfo[PK]): Boolean = {
+        assert(info.min >= max)
+        info.min == info.max && info.min == max
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: This is more like `overlapsAndSingleton`, right?

Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: Or maybe it should be:

```scala
def overlaps(info: PartitionKeyInfo[PK]): Boolean =
  info.min == max

def overlapsAndSingleton(info: PartitionKeyInfo[PK]): Boolean = {
  assert(info.min >= max)
  info.min == info.max && overlaps(info)
}
```

File: src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala [hunk 4]
```diff
@@ -40,7 +43,9 @@ class OrderedRDDSuite extends SparkSuite {
         Iterator((partitionIndex, sorted, correctPartitioning, if (a.nonEmpty) Some((a.head._1, a.last._1)) else None))
```
Comment (danking) [inline src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala]: Usually when I see a 4-tuple, I think it should become a class. If nothing more than to give names to fields.

```scala
case class PartitionSummary(partitionIndex: Int,
  sorted: Boolean, 
  correctPartitioning: Boolean,
  maybeBounds: Option[(Int, Int)])
```

**Pull request merged**